nemo/collections/speechlm2/models/salm.py

```diff
@@ -0,0 +1,588 @@
+        Notably:
+        * Convert source audio to speech representations.
+        * Convert target audio to target audio tokens.
+        * Convert target text to embeddings.
+        * Combine the input audio and target text embeddings.
+        * Take care of any necessary slicing to align the shapes of source audio,
+            target audio, and target token ids.
+        """
+        # Source audio encoding.
+        # Input audio: (B, T_samples)
+        # Audio embeddings: (B, T, H)
+        audio_embs, audio_emb_lens = self.perception(
+            input_signal=batch["audios"], input_signal_length=batch["audio_lens"]
+        )
+        audio_embs = [emb[:emblen] for emb, emblen in zip(audio_embs, audio_emb_lens)]
+        input_ids_to_embed = torch.where(batch["input_ids"] == self.audio_locator_tag_id, 0, batch["input_ids"])
+        text_embs = self.embed_tokens(input_ids_to_embed)
+        input_embs, target_ids, attention_mask = replace_placeholders_and_build_targets(
+            input_ids=batch["input_ids"],
+            embeds=text_embs,
+            padding_id=self.text_pad_id,
+            placeholder_id=self.audio_locator_tag_id,
+            replacements=audio_embs,
+            target_ids=batch["input_ids"].where(batch["loss_mask"], -100),
+        )
+        input_embs = input_embs[:, :-1]
+        attention_mask = attention_mask[:, :-1]
+        target_ids = target_ids[:, 1:]
+
+        # Combine target audio and text into a single tensor to slice them together.
+        # It will also help us truncate the sequence lengths to be divisible by TP world size,
+        # when TP is enabled.
+        # Input ids: (B, T, K+1)
+        if self._use_tp:
+            tp_world_size = self.device_mesh["tensor_parallel"].size()
+            if (remainder := (input_embs.shape[1] - 1) % tp_world_size) != 0:
+                # Truncate some tokens from the end to make the sequence lenght shape divisible by tensor parallelism
+                # world size. Otherwise, sequence parallelism will change the input shape making leading to mismatches.
+                input_embs = input_embs[:, :-remainder]
+                attention_mask = attention_mask[:, :-remainder]
+                target_ids = target_ids[:, :-remainder]
+
+        return {
+            "audio_embeds": audio_embs,
+            "text_embeds": text_embs,
+            "input_embeds": input_embs,
+            "attention_mask": attention_mask,
+            "target_ids": target_ids,
+        }
+
+    def training_step(self, batch: dict, batch_idx: int):
+        for m in (self.perception.preprocessor, self.perception.encoder, self.llm):
+            if is_frozen(m):
+                m.eval()
+
+        inputs = self.prepare_inputs(batch)
+        forward_outputs = self(inputs["input_embeds"], attention_mask=inputs["attention_mask"])
+        num_frames = (inputs["target_ids"] != -100).long().sum()
```

Comment:
why is there a constant `-100`?

Reply:
I added a comment: `# CrossEntropyLoss().ignore_index`

Reply:
Where can I find the comment? How is the value determined and will this value change in the future?